From 0589f69702fc2da8afe9a044a57137943b13f1cd Mon Sep 17 00:00:00 2001
From: Yiwei Zhang <zzyiwei@chromium.org>
Date: Thu, 13 Nov 2025 21:35:41 -0800
Subject: [PATCH] venus: avoid re-imported dma-buf to have a larger map size

If the allocation originates from the same instance, the tracker map
size follows the allocationSize. After export and re-import, mapping the
whole dma-buf can exceed the original map size. This change backs out
the offending changes.

Test: dEQP-VK.api.external.memory.*.suballocated.host_visible.*
Fixes: 442f242a495 ("venus: requests whole blob mem size for non-dedicated import")
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38443>
(cherry picked from commit c259ea24eea24868cfd0556316e1daaec6277321)
---
 .pick_status.json                       | 2 +-
 src/virtio/vulkan/vn_device_memory.c    | 9 ++-------
 src/virtio/vulkan/vn_renderer_virtgpu.c | 4 +---
 3 files changed, 4 insertions(+), 11 deletions(-)

diff --git a/.pick_status.json b/.pick_status.json
index 145dbc22b592d..78d420590ffa3 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -424,7 +424,7 @@
         "description": "venus: avoid re-imported dma-buf to have a larger map size",
         "nominated": true,
         "nomination_type": 2,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": "442f242a4953a10c558d9628eaadb79117648a23",
         "notes": null
diff --git a/src/virtio/vulkan/vn_device_memory.c b/src/virtio/vulkan/vn_device_memory.c
index 2795ea41c1d3f..8e362deeebf40 100644
--- a/src/virtio/vulkan/vn_device_memory.c
+++ b/src/virtio/vulkan/vn_device_memory.c
@@ -115,16 +115,11 @@ vn_device_memory_import_dma_buf(struct vn_device *dev,
    const VkMemoryType *mem_type =
       &dev->physical_device->memory_properties
           .memoryTypes[alloc_info->memoryTypeIndex];
-   const VkMemoryDedicatedAllocateInfo *dedicated_info =
-      vk_find_struct_const(alloc_info->pNext, MEMORY_DEDICATED_ALLOCATE_INFO);
-   const bool is_dedicated =
-      dedicated_info && (dedicated_info->image != VK_NULL_HANDLE ||
-                         dedicated_info->buffer != VK_NULL_HANDLE);
 
    struct vn_renderer_bo *bo;
    VkResult result = vn_renderer_bo_create_from_dma_buf(
-      dev->renderer, is_dedicated ? alloc_info->allocationSize : 0, fd,
-      mem_type->propertyFlags, &bo);
+      dev->renderer, alloc_info->allocationSize, fd, mem_type->propertyFlags,
+      &bo);
    if (result != VK_SUCCESS)
       return result;
 
diff --git a/src/virtio/vulkan/vn_renderer_virtgpu.c b/src/virtio/vulkan/vn_renderer_virtgpu.c
index 8f6d6ca8a5ce8..fe11f9f507cbe 100644
--- a/src/virtio/vulkan/vn_renderer_virtgpu.c
+++ b/src/virtio/vulkan/vn_renderer_virtgpu.c
@@ -1190,13 +1190,11 @@ virtgpu_bo_create_from_dma_buf(struct vn_renderer *renderer,
          /* If queried blob size is smaller than requested allocation size, we
           * drop the mappable flag to defer the mapping failure till the app's
           * vkMapMemory api call.
-          *
-          * Use size zero to request mapping the whole bo.
           */
          if (info.size < size)
             blob_flags &= ~VIRTGPU_BLOB_FLAG_USE_MAPPABLE;
          else
-            mmap_size = size > 0 ? size : info.size;
+            mmap_size = size;
       }
    }
 
-- 
GitLab

